conclave-enclave/src/main/kotlin/com/r3/conclave/enclave/internal/SocketEnclaveHostInterface.kt

```diff
@@ -122,41 +77,33 @@ class SocketEnclaveHostInterface(
             stateManager.transitionStateFrom<State.Running>(to = State.Stopped) {
                 "Call interface is not running."
             }
-            receiveLoopThread.join()
-            toHostSocket.close()
-            fromHostSocket.close()
+            callExecutor.shutdown()
         }
     }
 
-    /**
-     * Send a message to the receiving thread on the host side interface.
-     * Once received, these messages are routed to the appropriate call context in order to support concurrency.
-     * The synchronisation that occurs in this method (and the corresponding message on the host side) is the primary
-     * bottleneck of this call interface implementation.
-     */
-    private fun sendMessageToHost(message: SocketCallInterfaceMessage) {
-        val messageBytes = message.toByteArray()
-        synchronized(toHost) {
-            toHost.writeIntLengthPrefixBytes(messageBytes)
-            toHost.flush()
+    fun awaitTermination() {
+        var done = false
+        while(!done) {
+            try {
+                done = callExecutor.awaitTermination(10, TimeUnit.SECONDS)
```

Comment:
Delegate the timeout parameter to the caller. Use a single `Duration` parameter. This will avoid all this messy termination logic

Reply:
The timeout isn't important here. This function is just supposed to wait until the interface terminates, regardless of how long that takes.

Reply:
The implementation is not correct. You're not supposed to catch and ignore InterruptedException. Typically you rethrow into another exception, or more correctly you reinstate the interrupt flag for the current thread so that a higher caller can detect the interrupt and do something about it if they need to. Typically an interrupt is called to signal that the thread should stop executing.

If you simply delegate to `callExecutor.awaitTermination` with a user specified `Duration` then you avoid all this. Further, since this method is called from the main method, you can see quite clearly that there is no need to deal with interrupts, since it's the main thread executing. In fact a timeout duration doesn't even make sense in that context, which implies this method should have the timeout as an nullable parameter.

Reply:
Gotcha. 
I don't think there is an await termination without a timeout parameter (don't know why). I wouldn't have bothered with the loop in that case!

Reply:
There isn't actually, my bad! Just pass in the timeout as a param and let the main method set a timeout, something like max int seconcds or soemthing.